modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java

```diff
@@ -1549,70 +1547,201 @@ protected void insert(SimpleFeature feature, SimpleFeatureType featureType, Conn
-                
-                for (Iterator f = features.iterator(); f.hasNext();) {
-                    SimpleFeature feature = (SimpleFeature) f.next();
-                    
-                    List<Object> keyValues = null;
-                    boolean useExisting = Boolean.TRUE.equals(feature.getUserData().get(Hints.USE_PROVIDED_FID));
-                    if(useExisting) {
-                        keyValues = decodeFID(key, feature.getID(), true);
-                    } else if (!postInsert) {
-                        keyValues = getNextValues( key, cx );
+            } catch (SQLException e) {
+                String msg = "Error inserting features";
+                throw (IOException) new IOException(msg).initCause(e);
+            }
+        }
+    }
+
+    /**
+     * Specialized insertion for dialects that are using prepared statements.
+     */
+    private void insertPS(Collection<SimpleFeature> features, InsertionClassifier kind,
+                          SimpleFeatureType featureType, Connection cx, PrimaryKey key)
+            throws IOException, SQLException {
+        final PreparedStatementSQLDialect dialect = (PreparedStatementSQLDialect) getSQLDialect();
+
+        final KeysFetcher keysFetcher = KeysFetcher.create(this, cx,kind.useExisting, key);
+
+        final String sql = buildInsertPS(kind, featureType, keysFetcher, dialect);
+        LOGGER.log(Level.FINE, "Inserting new features with ps: {0}", sql);
+
+        //create the prepared statement
+        final PreparedStatement ps;
+        if (keysFetcher.isPostInsert()) {
+            // ask the DB to return the values of all the keys after the insertion
+            ps = cx.prepareStatement(sql, keysFetcher.getColumnNames());
+        } else {
+            ps = cx.prepareStatement(sql);
+        }
+        try {
+            for (SimpleFeature feature : features) {
+                //set the attribute values
+                int i = 1;
+                for(AttributeDescriptor att : featureType.getAttributeDescriptors()) {
+                    String colName = att.getLocalName();
+                    // skip the pk columns in case we have exposed them, we grab the
+                    // value from the pk itself
+                    if(keysFetcher.isKey(colName)) {
+                        continue;
                     }
-                    
 
-                    if ( dialect instanceof PreparedStatementSQLDialect ) {
-                        PreparedStatement ps = insertSQLPS( featureType, feature, keyValues, cx );
-                        try {
-                            ((PreparedStatementSQLDialect)dialect).onInsert(ps, cx, featureType);
-                            ps.execute();
-                        } finally {
-                            closeSafe( ps );
+                    Class binding = att.getType().getBinding();
+
+                    Object value = feature.getAttribute(colName);
+                    /*if (value == null) {
```

Comment:
That is old code I've moved without touching it. I wanted to avoid changing 
things too much. But no problem, I'll uncomment it and see if I don't break 
anything.

How should I move forward with your comments? Amend the existing commits or 
add new commits?

Envoyé avec AquaMail pour Android
http://www.aqua-mail.com


Le 25 octobre 2015 3:07:34 PM Andrea Aime <notifications@github.com> a écrit :

>>
>> -                    if ( dialect instanceof PreparedStatementSQLDialect ) {
>> -                        PreparedStatement ps = insertSQLPS( featureType, 
>> feature, keyValues, cx );
>> -                        try {
>> -                            
>> ((PreparedStatementSQLDialect)dialect).onInsert(ps, cx, featureType);
>> -                            ps.execute();
>> -                        } finally {
>> -                            closeSafe( ps );
>> +                    Class binding = att.getType().getBinding();
>> +
>> +                    Object value = feature.getAttribute(colName);
>> +                    /*if (value == null) {
>
> Do or don't, there is no try (don't add commented out sections of code, 
> just uncomment and throw the exception
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/geotools/geotools/pull/1009/files#r42945122